### DISCUSSION TITLE
Request for DFPlayer Mini Driver/Library for ESP8266 with Moddable SDK

### DISCUSSION THREAD

**Discussion by akbaraditamasp** [Moddable-OpenSource/moddable#1441] [Q&A] [answered]
Hi, Moddable SDK community!

I’m working on a project with ESP8266 and the DFPlayer Mini module, aiming to play MP3 files stored on an SD card. I’ve tried implementing communication using bit-banged serial, but it’s been quite challenging and not reliable.

Before investing more time in a custom solution, I wanted to ask:

Is there an existing driver or library for DFPlayer Mini that works with the Moddable SDK?
If not, has anyone successfully interfaced with DFPlayer Mini and could share a working example?
A prebuilt driver or guidance on how to efficiently handle communication with this module in JavaScript (via Moddable SDK) would be immensely helpful.

Thank you for any pointers or support!

**Comment by phoddie** [comment]
Welcome!

I'm not familiar with the DFPlayer Mini module. You mention "bit-banged serial." Is there a reason that you are not just using the serial driver? The ECMA-419 Serial IO module is available for ESP8266. The [standard](https://419.ecma-international.org/#-10-io-classes-serial) itself is a little challenging to read, but there are a couple of examples in this repository that should get you started: [continuouswrite](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/io/serial/continuouswrite/main.js) and [echo](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/io/serial/echo/main.js). I'm happy to try to answer any questions you may have about that.

**Reply by akbaraditamasp** [reply]
I've tried using the serial driver, but I encountered a message saying the pin is already in use. After researching, I found that the pin is indeed used for debugging. I also tried changing the pin, but it still didn't work. Then, I attempted to implement bit-banged serial, but that didn't work either. I suspect the issue isn't just with the serial, but that there might be something wrong with how I created the DFPlayer Mini driver as well. So, I've decided to ask for help from others who might have worked on a similar project to provide a working example.

**Reply by phoddie** [reply]
There are two serial ports on an ESP8266. The first is GPIO 1 & 3. If JavaScript debugging is enabled, those are indeed in use. You can run a release build instead, which will free up the pins. Just drop the `-d` from the `mcconfig` command line and you'll have a release build.  The second serial port is transmit only, on GPIO 2. I don't know if the DF Player Mini interface requires RX as well.

**Reply by akbaraditamasp** [reply]
It's really difficult to do this without debugging, since I'm still developing the driver and need a lot of trial and error. Meanwhile, I'm just trying to translate the DFPlayerMini library from C++ to JavaScript, and I'm encountering some difficulties. I’m still hoping someone here can help.

**Reply by phoddie** [reply]
Agreed: debuggers are great. That's why we created a JavaScript debugger can even work with microcontrollers. Can you connect the TX and RX pins of the DF Player Mini to another device? For example, if you connect them to an ESP32, there are at least two fully functioning serial ports, so you can use one for the DF Player Mini and the other for debugging. Or an ESP32-S3 with USB, so you can debug over USB and use Serial for the DF Player Mini. The serial code will be the same, since the ECMA-419 standard is supported on both ESP32 and ESP8266.

**Reply by akbaraditamasp** [reply]
I finally decided to move the pins for the DFPlayer. Can I change the Serial to use specific pins? I didn't see any examples of this at all.

**Reply by phoddie** [reply]
ESP8266 has two fixed serial ports, as noted above:

> There are two serial ports on an ESP8266. The first is GPIO 1 & 3.... The second serial port is transmit only, on GPIO 2. 

The pins cannot be changed: they are fixed in hardware. Here's the pinout diagram that shows them.

![esp8266-Pinout-3351485625](https://github.com/user-attachments/assets/eac3e697-a8be-4f44-afba-add88add7347)

You select the transmit and receive pins in the constructor options object as documented in the [link to ECMA-419](https://419.ecma-international.org/#-10-io-classes-serial) above:

```js
let serial = new device.io.Serial({
	...device.Serial.default,
	transmit: 1,
	receive: 3,
	baud: 115200,
	format: "buffer",
	// etc...
});
```

**Reply by akbaraditamasp** [reply]
Is there no library to perform something like SoftwareSerial in Arduino?

**Reply by akbaraditamasp** [reply, marked as answer]
Finally, it works! It turns out that I needed to include `onReadable`. It seems to be required to forward commands sent to the Serial. Perhaps I wasn't careful enough when reading the documentation.

```javascript
const serial = new device.io.Serial({
  ...device.Serial.default,
  baud: 9600,
  format: "number",
  // This part solved my problem
  onReadable: function (count) {
    while (count--) this.write(this.read());
  },
});
```

I might release the DFPlayer library soon after this. Thank you!
